server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -0,0 +1,343 @@
+            return new MatchNoneQueryToFilterAdapter(searcher, key);
+        }
+        return new DefaultQueryToFilterAdapter(searcher, key, query);
+    }
+
+    private final IndexSearcher searcher;
+    private final String key;
+
+    private QueryToFilterAdapter(IndexSearcher searcher, String key) {
+        this.searcher = searcher;
+        this.key = key;
+    }
+
+    public final String key() {
+        return key;
+    }
+
+    protected final IndexSearcher searcher() {
+        return searcher;
+    }
+
+    /**
+     * Is it safe to use index metadata like
+     * {@link IndexReader#docFreq} or {@link IndexReader#maxDoc} to count. the
+     * number of matching documents.
+     */
+    protected final boolean countCanUseMetadata(FiltersAggregator.Counter counter, Bits live) {
+        return live == null && counter.docCount.alwaysOne();
+    }
+
+    /**
+     * Make a filter that matches this filter and the provided query.
+     * <p>
+     * Note: This method rewrites the query against the {@link IndexSearcher}.
+     */
+    abstract QueryToFilterAdapter union(Query extraQuery) throws IOException;
+
+    abstract IntPredicate matchingDocIds(LeafReaderContext ctx) throws IOException;
+
+    abstract long count(LeafReaderContext ctx, FiltersAggregator.Counter counter, Bits live) throws IOException;
+
+    /**
+     * Estimate the cost of calling {@code #count} on this leaf.
+     */
+    abstract long estimateCountCost(LeafReaderContext ctx, CheckedSupplier<Boolean, IOException> canUseMetadata) throws IOException;
+
+    /**
+     * Collect profiling information for this filter.
+     */
+    void collectDebugInfo(BiConsumer<String, Object> add) {}
+
+    /**
+     * Special case when the filter can't match anything.
+     */
+    private static class MatchNoneQueryToFilterAdapter extends QueryToFilterAdapter {
+        private MatchNoneQueryToFilterAdapter(IndexSearcher searcher, String key) {
+            super(searcher, key);
+        }
+
+        @Override
+        QueryToFilterAdapter union(Query extraQuery) throws IOException {
+            return this;
+        }
+
+        @Override
+        IntPredicate matchingDocIds(LeafReaderContext ctx) throws IOException {
+            return l -> false;
+        }
+
+        @Override
+        long count(LeafReaderContext ctx, FiltersAggregator.Counter counter, Bits live) throws IOException {
+            return 0;
+        }
+
+        @Override
+        long estimateCountCost(LeafReaderContext ctx, CheckedSupplier<Boolean, IOException> canUseMetadata) throws IOException {
+            return 0;
+        }
+
+        @Override
+        void collectDebugInfo(BiConsumer<String, Object> add) {
+            super.collectDebugInfo(add);
+            add.accept("type", "match_none");
+        }
+    }
+
+    /**
+     * Filter that matches every document.
+     */
+    private static class MatchAllQueryToFilterAdapter extends QueryToFilterAdapter {
```

Comment:
And you don't strictly need this one either because folks won't usually write `match_all` in their `filters` agg. But, again, its a fairly compact way of looking at how you'd read from metadata.